crates/types/src/config/mod.rs

```diff
@@ -259,18 +260,47 @@ impl Configuration {
             return Err(InvalidConfigurationError::ForceNodeIdZero);
         }
 
+        if self.common.node_name.is_none() {
+            // If the node name is not set, we will fallback to use hostname as the node name.
+            // So to avoid changing hostname to make data loss, we must validate the directory's entry.
+            let dirs = read_dir_subdirs(&self.common.base_dir());
+            match dirs.len().cmp(&1) {
+                std::cmp::Ordering::Less => {
+                    // If it's an empty directory, it's safe to use default behavior.
+                }
+                std::cmp::Ordering::Equal => {
+                    // If there is only one directory, it must be the node name.
+                    // And if it's not equal to hostname, return an error.
+                    if self.common.node_name() != dirs[0].to_string_lossy() {
+                        return Err(InvalidConfigurationError::RequiredNodeName(format!(
+                            "The working directory have only one subdirectory '{}', and node name is not set, but it's not equal to hostname '{}'",
+                            dirs[0].to_string_lossy(),
+                            self.common.node_name()
+                        )));
```

Comment:
As a follow-up, we could consider setting the value of a single sub dir as the node_name if it is unset in `set_derived_values`. But the check itself is already making things a lot less surprising :-)

Reply:
I prefer not to make any modifications to the values. Because in this situation (that is, when the nodename doesn't match the data in the directory), it's difficult to distinguish whether the user has modified the hostname or has mistakenly started the system in a directory that already contains data. It seems that asking the user to clearly indicate how to handle this situation might be a safer approach.